go/cmd/server/logging/logging.go

```diff
@@ -0,0 +1,359 @@
+		arrowType, err := gotypes.ValueTypeEnumToArrowType(fcoSchema.EntityTypes[entityName])
+		if err != nil {
+			return nil, err
+		}
+		fields = append(fields, arrow.Field{
+			Name: entityName,
+			Type: arrowType,
+		})
+		columns = append(columns, valArrowArray)
+	}
+
+	for _, featureName := range fcoSchema.Features {
+
+		protoArr := columnNameToProtoValueArray[featureName]
+		if len(protoArr) == 0 {
+			break
+		}
+		arrowArray, err := gotypes.ProtoValuesToArrowArray(protoArr, arrowMemory, len(columnNameToProtoValueArray))
+		if err != nil {
+			return nil, err
+		}
+
+		arrowType, err := gotypes.ValueTypeEnumToArrowType(fcoSchema.FeaturesTypes[featureName])
+
+		if err != nil {
+			return nil, err
+		}
+		fields = append(fields, arrow.Field{
+			Name: featureName,
+			Type: arrowType,
+		})
+		columns = append(columns, arrowArray)
+	}
+	schema := arrow.NewSchema(
+		fields,
+		nil,
+	)
+
+	result := array.Record(array.NewRecord(schema, columns, int64(len(memoryBuffer.logs))))
+
+	tbl := array.NewTableFromRecords(schema, []array.Record{result})
+	return array.Table(tbl), nil
+}
+
+type Schema struct {
+	Entities      []string
+	Features      []string
+	EntityTypes   map[string]types.ValueType_Enum
+	FeaturesTypes map[string]types.ValueType_Enum
+}
+
+func GetSchemaFromFeatureService(featureService *model.FeatureService, entities []*model.Entity, featureViews []*model.FeatureView, onDemandFeatureViews []*model.OnDemandFeatureView) (*Schema, error) {
+	fvs := make(map[string]*model.FeatureView)
+	odFvs := make(map[string]*model.OnDemandFeatureView)
+
+	joinKeys := make([]string, 0)
+	for _, entity := range entities {
+		joinKeys = append(joinKeys, entity.JoinKey)
+	}
+
+	for _, featureView := range featureViews {
+		fvs[featureView.Base.Name] = featureView
+	}
+
+	for _, onDemandFeatureView := range onDemandFeatureViews {
+		odFvs[onDemandFeatureView.Base.Name] = onDemandFeatureView
+	}
+
+	entityJoinKeyToType := make(map[string]types.ValueType_Enum)
+	for _, entity := range entities {
+		entityJoinKeyToType[entity.JoinKey] = entity.ValueType
+	}
+
+	allFeatureTypes := make(map[string]types.ValueType_Enum)
+	features := make([]string, 0)
+	for _, featureProjection := range featureService.Projections {
+		// Create copies of FeatureView that may contains the same *FeatureView but
+		// each differentiated by a *FeatureViewProjection
+		featureViewName := featureProjection.Name
+		if fv, ok := fvs[featureViewName]; ok {
+			for _, f := range fv.Base.Projection.Features {
+				// add feature to map
+				features = append(features, f.Name)
+				allFeatureTypes[f.Name] = f.Dtype
+			}
+		} else if odfv, ok := odFvs[featureViewName]; ok {
+			for name, valueType := range odfv.GetRequestDataSchema() {
+				features = append(features, name)
+				allFeatureTypes[name] = valueType
+			}
+		} else {
+			return nil, fmt.Errorf("no such feature view found in feature service %s", featureViewName)
+		}
+	}
+	schema := &Schema{
+		Entities:      joinKeys,
```

Comment:
I guess here we need only joinKeys that required by current feature service

Reply:
For this, there no way of actually loading entities based on feature views or feature service. Should I add this functionality? I will have to update either featureservice to contain references to each entitiy or the entitys that reference the featureviews.

Reply:
```
joinKeysSet = make(map[string]interface{})

for _, projection := range featureService.Projections {
    fv := fvs[projection.Name]
    for _, entityName := range fv.Entities {
          entity := entities[entityName]
          if joinKeyAlias, ok := projection.JoinKeyMap[entity.JoinKey]; ok {
            joinKeysSet[joinKeyAlias] = nil
          } else {
            joinKeysSet[entity.JoinKey] = nil
          }
    }
}
```

Reply:
Done.